File: may2020/7-convert-binary-number-in-a-linked-list-to-integer/solution.scala
/**
 * Definition for singly-linked list.
 * class ListNode(_x: Int = 0, _next: ListNode = null) {
 *   var next: ListNode = _next
 *   var x: Int = _x
 * }
 */
object Solution {
    def getDecimalValue(head: ListNode): Int = {
        def doGetDecimalValue(result: Int, head: ListNode): Int = {
          if (head == null) result else doGetDecimalValue(result * 2 + head.x, head.next)
        }
        doGetDecimalValue(0, head)         
        //if (head == null) 0 else (head.x * 2) + getDecimalValue(head.next)
    }
}

/*
 * from https://leetcode.com/problems/convert-binary-number-in-a-linked-list-to-integer/
 * Input: [1, 0, 1]
 * Output: 5
 * Expected: 5
 */

File: may2020/7-convert-binary-number-in-a-linked-list-to-integer/solution.swift
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func getDecimalValue(_ head: ListNode?) -> Int {
        var result = 0
        var node = head
        while (node != nil) {
            result = result * 2 + (node?.val ?? 0)
            node = node?.next
        }
        return result
    }
}

/*
 * from https://leetcode.com/problems/convert-binary-number-in-a-linked-list-to-integer/
 * Input: [1, 0, 1]
 * Output: 5
 * Expected: 5
 */
